### DISCUSSION TITLE
Move project

### DISCUSSION THREAD

**Discussion by ben-hamida** [VsixCommunity/Community.VisualStudio.Toolkit#266] [Q&A]
I want to be able to move a project from one directory to another while keeping the project path in the solution file up to date.

I tried manually moving the project on disk and then patching the solution file but that displays the File Modification Detected dialog, and that I want to avoid.

I tried implementing [IVsPersistSolutionProps](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivspersistsolutionprops?redirectedfrom=MSDN&view=visualstudiosdk-2022) but I found no way of writing the project path.

After some searching I found the [IVsSolution3.UpdateProjectFileLocationForUpgrade](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution3.updateprojectfilelocationforupgrade?view=visualstudiosdk-2022) but it didn't seem to do anything. Same with [IVsSolution.OnAfterRenameProject](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution.onafterrenameproject?view=visualstudiosdk-2022).

The last thing I tried was to remove the project with [IVsSolution.CloseSolutionElement](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolution.closesolutionelement?view=visualstudiosdk-2022#microsoft-visualstudio-shell-interop-ivssolution-closesolutionelement(system-uint32-microsoft-visualstudio-shell-interop-ivshierarchy-system-uint32)) and then readding it with the new path using [_Solution.AddFromFile](https://docs.microsoft.com/en-us/dotnet/api/envdte._solution.addfromfile?view=visualstudiosdk-2022#envdte-solution-addfromfile(system-string-system-boolean)), but the AddFromFile call threw an InvalidOperationException with the message 'This operation is not valid while a project lock is held.'. This solution doesn't feel right anyway, there could be project specific properties in the solution file that wouldn't be preserved.

Is there a way to update the project path in the solution?

**Comment by reduckted** [comment]
I think you were on the right track with `IVsSolution.CloseSolutionElement`. I've been able to move a project on disk by doing this:

```cs
IVsSolution solution = await VS.Services.GetSolutionAsync();
Project projectToMove = GetProjectToMove(); // Find the project that you want to move.

// Remember the old path and project file name.
string oldPath = Path.GetDirectoryName(projectToMove.FullPath);
string name = Path.GetFileName(projectToMove.FullPath);

// Remove the project from the solution.
projectToMove.GetItemInfo(out IVsHierarchy hierarchy, out _, out _);
solution.CloseSolutionElement((uint)__VSSLNCLOSEOPTIONS.SLNCLOSEOPT_DeleteProject, hierarchy, 0);

// Move the project directory on disk.
string newPath = oldPath + "_moved";
Directory.Move(oldPath, newPath);

// Add the existing project to the solution.
solution.CreateProject(
    default,
    Path.Combine(newPath, name), // Specifying a file name here and null for the next two parameters will add an existing project.
    null,
    null,
    (uint)__VSCREATEPROJFLAGS.CPF_OPENFILE,
    typeof(IVsHierarchy).GUID,
    out IntPtr projectPtr
);

try
{
    // Get the new IVsHierachy from the pointer.
    IVsHierarchy newHierarchy = (IVsHierarchy)Marshal.GetUniqueObjectForIUnknown(projectPtr);
    Project newProject = (Project)await SolutionItem.FromHierarchyAsync(newHierarchy, VSConstants.VSITEMID_ROOT);

    // Do something with the new project.
    await VS.MessageBox.ShowAsync(newProject.FullPath);
}
finally
{
    Marshal.Release(projectPtr);
}
```

Two things that might be problems for you:
1. The project is added directly under the solution. I couldn't find a way to add it to a solution folder, but it must be possible somehow.
2. You will need to find a way to update any project references.

**Reply by ben-hamida** [reply]
Thanks! It seems backward to have to remove and readd a project to change its path. It's unfortunate that there is no API to set the path, seems like such a simple thing.

The problems you mention sound very solvable. My bigger concern is the preserving of any project related properties that might exist in the solution file. Say for example that there is a custom build configuration in addition to the standard Debug and Release, that would probably not be preserved. Although probably in 99% of the cases it wouldn't be an issue. I also wonder if it's important to preserve the GUID of the project, but I'm guessing it's not possible to set that either.

Another idea I have is to programmatically make sure the solution is saved, then patch the solution file on disk and then immediate reload the solution to avoid the dialog about solution having been modified outside the environment. Haven't tried it yet and I'm not too optimistic that it will work. I have to find the right API to reload or refresh the solution first.
